client/gossipEndpointDiscoverer.go

```diff
@@ -1,84 +1,133 @@
 package client
 
 import (
-	"crypto/tls"
-	"encoding/json"
-	"errors"
+	"context"
 	"fmt"
-	"io/ioutil"
 	"log"
 	"math/rand"
-	"net/http"
-	"net/url"
-)
+	"time"
 
-// GossipEndpointDiscoverer used for discovering and picking the most appropriate node in a cluster
-type GossipEndpointDiscoverer struct {
-	MaxDiscoverAttempts       int
-	GossipSeeds               []*url.URL
-	NodePreference            NodePreference
-	SkipCertificateValidation bool
-	httpClient                *http.Client
-}
+	gossipApi "github.com/EventStore/EventStore-Client-Go/protos/gossip"
+	"github.com/EventStore/EventStore-Client-Go/protos/shared"
+	"google.golang.org/grpc"
+)
 
-func NewGossipEndpointDiscoverer() *GossipEndpointDiscoverer {
-	transport := &http.Transport{
-		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
-	}
-	client := &http.Client{Transport: transport}
-	return &GossipEndpointDiscoverer{
-		MaxDiscoverAttempts:       10,
-		NodePreference:            NodePreference_Leader,
-		SkipCertificateValidation: false,
-		httpClient:                client,
+func allowedNodeState() []gossipApi.MemberInfo_VNodeState {
+	return []gossipApi.MemberInfo_VNodeState{
+		gossipApi.MemberInfo_Follower,
+		gossipApi.MemberInfo_Leader,
+		gossipApi.MemberInfo_ReadOnlyLeaderless,
+		gossipApi.MemberInfo_PreReadOnlyReplica,
 	}
 }
 
-// Discover will discover nodes via performing a gossip over HTTP and then picking the best candidate to connect to
-func (discoverer *GossipEndpointDiscoverer) Discover() (*MemberInfo, error) {
-	if len(discoverer.GossipSeeds) == 0 {
-		return nil, errors.New("There are no gossip seeds")
-	}
-	discoverer.GossipSeeds = shuffleGossipSeeds(discoverer.GossipSeeds)
-	gossipIndex := 0
-	for attempt := 1; attempt <= discoverer.MaxDiscoverAttempts; attempt++ {
-		if gossipIndex >= len(discoverer.GossipSeeds) {
-			gossipIndex = 0
+func DiscoverNode(conf *Configuration) (*grpc.ClientConn, error) {
+	var connection *grpc.ClientConn = nil
+	attempt := 1
+
+	if conf.DnsDiscover || len(conf.GossipSeeds) > 0 {
+		var candidates []string
+
+		if conf.DnsDiscover {
+			candidates = append(candidates, conf.Address)
+		} else {
+			for _, seed := range conf.GossipSeeds {
+				candidates = append(candidates, seed.Host)
```

Comment:
```suggestion
				candidates = append(candidates, seed.String())
```
Otherwise the port gets lost